src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,551 @@
+	return sqrt_int16((uint16_t)xs);
+}
+
+static inline int16_t max_mic_distance(struct tdfb_comp_data *cd)
+{
+	int16_t d;
+	int16_t dx;
+	int16_t dy;
+	int16_t dz;
+	int32_t d2;
+	int32_t d2_max = 0;
+	int i;
+	int j;
+
+	/* Max lag based on largest array dimension. Microphone coordinates are Q4.12 meters */
+	for (i = 0; i < cd->config->num_mic_locations; i++) {
+		for (j = 0; i < cd->config->num_mic_locations; i++) {
+			if (j == i)
+				continue;
+
+			dx = cd->mic_locations[i].x - cd->mic_locations[j].x;
+			dy = cd->mic_locations[i].y - cd->mic_locations[j].y;
+			dz = cd->mic_locations[i].z - cd->mic_locations[j].z;
+
+			/* d2 is Q8.24 meters */
+			d2 = dx * dx + dy * dy + dz * dz;
+			if (d2 > d2_max)
+				d2_max = d2;
+		}
+	}
+
+	/* Squared distance is Q8.24, return Q4.12 meters */
+	d = tdfb_mic_distance_sqrt(d2_max);
+	return d;
+}
+
+static inline bool line_array_mode_check(struct tdfb_comp_data *cd)
+{
+	int32_t px, py, pz;
+	int16_t a, b, c, d, e, f;
+	int i;
+	int num_mic_locations = cd->config->num_mic_locations;
+
+	if (num_mic_locations == 2)
+		return true;
+
+	/* Cross product of vectors AB and AC is (0, 0, 0) if they are co-linear.
+	 * Form vector AB(a,b,c) from x(i+1) - x(i), y(i+1) - y(i), z(i+1) - z(i)
+	 * Form vector AC(d,e,f) from x(i+2) - x(i), y(i+2) - y(1), z(i+2) - z(i)
+	 * Cross product AB x AC is (b * f - c * e, c * d - a * f, a * e - b * d)
+	 */
+	for (i = 0; i < num_mic_locations - 3; i++) {
+		a = cd->mic_locations[i + 1].x - cd->mic_locations[i].x;
+		b = cd->mic_locations[i + 1].y - cd->mic_locations[i].y;
+		c = cd->mic_locations[i + 1].z - cd->mic_locations[i].z;
+		d = cd->mic_locations[i + 2].x - cd->mic_locations[i].x;
+		e = cd->mic_locations[i + 2].y - cd->mic_locations[i].y;
+		f = cd->mic_locations[i + 2].z - cd->mic_locations[i].z;
+		px = (int32_t)b * f - (int32_t)c * e;
+		py = (int32_t)c * d - (int32_t)a * f;
+		pz = (int32_t)a * e - (int32_t)b * d;
+		if (px || py || pz)
+			return false;
+	}
+
+	return true;
+}
+
+int tdfb_direction_init(struct tdfb_comp_data *cd, int32_t fs, int ch_count)
+{
+	struct sof_eq_iir_header_df2t *filt;
+	int64_t *delay;
+	int32_t d_max;
+	int32_t t_max;
+	size_t size;
+	int n;
+	int i;
+
+	/* Select emphasis response per sample rate */
+	switch (fs) {
+	case 16000:
+		filt = (struct sof_eq_iir_header_df2t *)iir_emphasis_16k;
+		break;
+	case 48000:
+		filt = (struct sof_eq_iir_header_df2t *)iir_emphasis_48k;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* Allocate delay lines for IIR filters and initialize them */
+	size = ch_count * iir_delay_size_df2t(filt);
+	delay = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, size);
+	if (!delay)
+		return -EINVAL;
+
+	cd->direction.df2t_delay = delay;
+	for (i = 0; i < ch_count; i++) {
+		iir_init_coef_df2t(&cd->direction.emphasis[i], filt);
+		iir_init_delay_df2t(&cd->direction.emphasis[i], &delay);
+	}
+
+	/* Unit delay length in Q1.31 seconds */
+	cd->direction.unit_delay = ((int64_t)1 << 31) / fs;
```

Comment:
isnt this just BIT(31)?

Reply:
It is, but a different context. Q_SHIFT_LEFT((int64_t)1, 0, 31) perhaps?